File: puzzle_selector.py
import random
import logging
from typing import Dict, List, Union, TypedDict, Optional
from dataclasses import dataclass

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Constants
ALPHA = 0.8  # Weight for theme selection based on rating differences
BETA = 1.0   # Weight for puzzle popularity in scoring
GAMMA = 1.0  # Weight for rating difference impact in scoring

class Puzzle(TypedDict):
    puzzleID: str
    lichessRating: int
    popularity: float
    themes: List[str]
    used: bool

def compute_theme_probabilities(user_theme_ratings: Dict[str, int]) -> Dict[str, float]:
    """
    Compute probability distribution over themes based on user ratings.
    
    Args:
        user_theme_ratings: Dictionary mapping themes to user's rating in that theme
    
    Returns:
        Dictionary mapping themes to their selection probabilities
    """
    max_rating = max(user_theme_ratings.values())
    logger.info(f"Computing theme probabilities. Max rating: {max_rating}")
    
    # Compute raw scores
    theme_scores = {
        theme: ALPHA * ((max_rating - rating) + 1) + (1 - ALPHA) * 1
        for theme, rating in user_theme_ratings.items()
    }
    
    # Normalize to probabilities
    total_score = sum(theme_scores.values())
    theme_probs = {theme: score/total_score for theme, score in theme_scores.items()}
    
    logger.info("Theme probabilities:")
    for theme, prob in theme_probs.items():
        logger.info(f"  {theme}: {prob:.3f} (rating: {user_theme_ratings[theme]})")
    
    return theme_probs

def pick_theme(theme_probs: Dict[str, float]) -> str:
    """
    Randomly select a theme based on provided probabilities.
    
    Args:
        theme_probs: Dictionary mapping themes to their selection probabilities
    
    Returns:
        Selected theme
    """
    themes = list(theme_probs.keys())
    probabilities = list(theme_probs.values())
    selected = random.choices(themes, weights=probabilities, k=1)[0]
    logger.info(f"Selected theme: {selected} (probability: {theme_probs[selected]:.3f})")
    return selected

def compute_puzzle_score(
    puzzle: Puzzle,
    user_theme_ratings: Dict[str, int],
    beta: float = BETA,
    gamma: float = GAMMA
) -> float:
    """
    Compute score for a puzzle based on user ratings and puzzle properties.
    """
    # Compute average user rating for puzzle themes
    theme_ratings = [user_theme_ratings[t] for t in puzzle['themes'] if t in user_theme_ratings]
    if not theme_ratings:
        logger.warning(f"Puzzle {puzzle['puzzleID']} has no valid themes for user ratings")
        return 0.0
    
    puzzle_rating_for_user = sum(theme_ratings) / len(theme_ratings)
    diff = abs(puzzle['lichessRating'] - puzzle_rating_for_user)
    
    # Compute components of the score
    popularity_component = puzzle['popularity'] ** beta
    rating_penalty = 1 / (1 + diff ** gamma)
    score = popularity_component * rating_penalty
    
    logger.info(f"""
Puzzle {puzzle['puzzleID']} score calculation:
- Puzzle rating: {puzzle['lichessRating']}
- Themes: {puzzle['themes']}
- Theme ratings used: {theme_ratings}
- User's average theme rating: {puzzle_rating_for_user:.1f}
- Rating difference: {diff}
- Popularity: {puzzle['popularity']:.2f}
- Popularity component (pop^{beta}): {popularity_component:.3f}
- Rating penalty (1/(1+diff^{gamma})): {rating_penalty:.3f}
- Final score: {score:.3f}
""")
    
    return score

def pick_puzzle_for_theme(
    theme: str,
    puzzles: List[Puzzle],
    user_theme_ratings: Dict[str, int]
) -> Optional[Puzzle]:
    """
    Pick a puzzle for a given theme based on user ratings.
    """
    # Filter puzzles by theme and unused status
    valid_puzzles = [p for p in puzzles if not p['used'] and theme in p['themes']]
    
    logger.info(f"\n=== Picking puzzle for theme: {theme} ===")
    logger.info(f"User rating for theme: {user_theme_ratings.get(theme)}")
    logger.info(f"Found {len(valid_puzzles)} valid puzzles for theme {theme}")
    
    if not valid_puzzles:
        logger.warning(f"No valid puzzles found for theme {theme}")
        return None
    
    # Log distribution of available puzzles
    ratings = [p['lichessRating'] for p in valid_puzzles]
    logger.info(f"""
Available puzzle statistics:
- Count: {len(valid_puzzles)}
- Rating range: {min(ratings)} to {max(ratings)}
- Available puzzles:""")
    
    # Compute scores for all valid puzzles
    puzzle_scores = []
    for puzzle in valid_puzzles:
        score = compute_puzzle_score(puzzle, user_theme_ratings)
        puzzle_scores.append(score)
    
    # Select puzzle weighted by scores
    selected_puzzle = random.choices(valid_puzzles, weights=puzzle_scores, k=1)[0]
    selected_puzzle['used'] = True
    
    logger.info("\n=== Final Selection ===")
    logger.info(f"""Selected puzzle {selected_puzzle['puzzleID']}:
- Rating: {selected_puzzle['lichessRating']}
- Themes: {selected_puzzle['themes']}
- Popularity: {selected_puzzle['popularity']:.2f}
- Score: {puzzle_scores[valid_puzzles.index(selected_puzzle)]:.3f}
- Rating difference from user: {abs(selected_puzzle['lichessRating'] - user_theme_ratings.get(theme, 0))}
""")
    
    return selected_puzzle

def get_next_puzzle(
    user_theme_ratings: Dict[str, int],
    puzzles: List[Puzzle]
) -> Optional[Puzzle]:
    """
    Get the next puzzle for a user based on their theme ratings.
    
    Args:
        user_theme_ratings: Dictionary of user's ratings in different themes
        puzzles: List of available puzzles
    
    Returns:
        Selected puzzle or None if no suitable puzzle found
    """
    logger.info("Starting puzzle selection process")
    logger.info(f"User has ratings for {len(user_theme_ratings)} themes")
    
    # Compute theme probabilities and select theme
    theme_probs = compute_theme_probabilities(user_theme_ratings)
    selected_theme = pick_theme(theme_probs)
    
    # Pick and return puzzle for selected theme
    return pick_puzzle_for_theme(selected_theme, puzzles, user_theme_ratings) 

def test_puzzle_selection():
    """
    Test function to debug puzzle selection with a specific case.
    """
    # User ratings from the image
    user_theme_ratings = {
        "capturingDefender": 1666,
        "discoveredAttack": 1398,
        "deflection": 1333,
        "quietMove": 1180,
        "fork": 1200,
        "kingsideAttack": 1200
    }
    
    # Create test puzzles including the one from the image
    puzzles = [
        {
            "puzzleID": "AMzCl",
            "lichessRating": 2354,
            "popularity": 1.0,  # 100% popularity
            "themes": ["kingsideAttack"],
            "used": False
        },
        # Add some other puzzles for comparison
        {
            "puzzleID": "test1",
            "lichessRating": 1250,
            "popularity": 0.8,
            "themes": ["kingsideAttack"],
            "used": False
        },
        {
            "puzzleID": "test2",
            "lichessRating": 1500,
            "popularity": 0.6,
            "themes": ["kingsideAttack", "fork"],
            "used": False
        }
    ]
    
    logger.info("=== Starting Test Puzzle Selection ===")
    
    # First, let's see the theme probabilities
    theme_probs = compute_theme_probabilities(user_theme_ratings)
    
    # Now let's force select kingsideAttack theme to match our case
    logger.info("\n=== Testing puzzle selection for kingsideAttack theme ===")
    puzzle = pick_puzzle_for_theme("kingsideAttack", puzzles, user_theme_ratings)
    
    if puzzle:
        logger.info("\n=== Final Selected Puzzle ===")
        logger.info(f"PuzzleID: {puzzle['puzzleID']}")
        logger.info(f"Rating: {puzzle['lichessRating']}")
        logger.info(f"Themes: {puzzle['themes']}")
        logger.info(f"Popularity: {puzzle['popularity']}")
    
    return puzzle

if __name__ == "__main__":
    test_puzzle_selection() 

File: src/utils/puzzleSelector.ts
import { Puzzle } from '../types/puzzle';

// Constants
const ALPHA = 0.8;  // Weight for theme selection based on rating differences
const BETA = 1.0;   // Weight for puzzle popularity in scoring
const GAMMA = 1.0;  // Weight for rating difference impact in scoring
const DEFAULT_RATING = 1200;  // Default rating for new themes

// Update rating range to be tighter
const RATING_RANGE = 150; // Look for puzzles within ±150 of user's rating
const MAX_PUZZLES_TO_SCORE = 50; // Maximum number of puzzles to score for efficiency

interface UserThemeRatings {
    [theme: string]: number;
}

function log(message: string, data?: any) {
    console.log(`[PuzzleSelector] ${message}`, data ? data : '');
}

function computeThemeProbabilities(userThemeRatings: UserThemeRatings): { [theme: string]: number } {
    // If no ratings exist, create default ratings for all themes
    if (Object.keys(userThemeRatings).length === 0) {
        log('No theme ratings found. Using equal probabilities for all themes.');
        return {
            'Mate': 0.167,
            'Endgame': 0.167,
            'Defense': 0.167,
            'Quiet Move': 0.167,
            'Capturing Defender': 0.167,
            'Fork': 0.167
        };
    }

    const maxRating = Math.max(...Object.values(userThemeRatings));
    log(`Computing theme probabilities. Max rating: ${maxRating}`);
    
    // Compute raw scores - higher score means MORE likely to be selected
    const themeScores: { [theme: string]: number } = {};
    
    // First, ensure all themes have a base score
    const allThemes = ['Mate', 'Endgame', 'Defense', 'Quiet Move', 'Capturing Defender', 'Fork'];
    allThemes.forEach(theme => {
        if (theme in userThemeRatings) {
            // For rated themes, prefer those with lower ratings
            themeScores[theme] = ALPHA * ((maxRating - userThemeRatings[theme]) + 1) + (1 - ALPHA);
        } else {
            // For unrated themes, give them a slightly higher chance to encourage exploration
            themeScores[theme] = ALPHA * ((maxRating - DEFAULT_RATING) + 1) + (1 - ALPHA) * 1.2;
        }
    });
    
    // Normalize to probabilities
    const totalScore = Object.values(themeScores).reduce((a, b) => a + b, 0);
    const themeProbs: { [theme: string]: number } = {};
    for (const [theme, score] of Object.entries(themeScores)) {
        themeProbs[theme] = score / totalScore;
    }
    
    log('Theme probabilities:');
    Object.entries(themeProbs).forEach(([theme, prob]) => {
        log(`  ${theme}: ${prob.toFixed(3)} (rating: ${userThemeRatings[theme] || DEFAULT_RATING}${theme in userThemeRatings ? '' : ' default'})`);
    });
    
    return themeProbs;
}

function pickTheme(themeProbs: { [theme: string]: number }): string {
    const themes = Object.keys(themeProbs);
    if (themes.length === 0) {
        log('No themes available. Using default theme.');
        return 'Mate';  // Changed default to Mate as it's more fundamental
    }

    // Use weighted random selection
    const rand = Math.random();
    let cumSum = 0;
    
    // Sort themes by probability for more predictable selection
    const sortedThemes = themes.sort((a, b) => themeProbs[b] - themeProbs[a]);
    
    for (const theme of sortedThemes) {
        cumSum += themeProbs[theme];
        if (rand <= cumSum) {
            log(`Selected theme: ${theme} (probability: ${themeProbs[theme].toFixed(3)})`);
            return theme;
        }
    }
    
    // Fallback to highest probability theme
    const selectedTheme = sortedThemes[0];
    log(`Selected theme (fallback): ${selectedTheme} (probability: ${themeProbs[selectedTheme].toFixed(3)})`);
    return selectedTheme;
}

function computePuzzleScore(
    puzzle: Puzzle,
    userThemeRatings: UserThemeRatings,
    beta: number = BETA,
    gamma: number = GAMMA
): number {
    // Compute average user rating for puzzle themes
    const themeRatings = puzzle.themes
        .map(t => userThemeRatings[t] || DEFAULT_RATING);  // Use default rating for unknown themes
    
    if (themeRatings.length === 0) {
        log(`Puzzle ${puzzle.id} has no valid themes`);
        return 0.0;
    }
    
    const puzzleRatingForUser = themeRatings.reduce((a, b) => a + b, 0) / themeRatings.length;
    const diff = Math.abs(puzzle.rating - puzzleRatingForUser);
    
    // Compute components of the score
    const popularityComponent = Math.pow(puzzle.popularity / 100, beta);  // Divide by 100 as popularity is in percentage
    const ratingPenalty = 1 / (1 + Math.pow(diff, gamma));
    const score = popularityComponent * ratingPenalty;
    
    log(`
Puzzle ${puzzle.id} score calculation:
- Puzzle rating: ${puzzle.rating}
- Themes: ${puzzle.themes.join(', ')}
- Theme ratings: ${themeRatings.join(', ')} (weighted average: ${puzzleRatingForUser.toFixed(1)})
- Rating difference: ${diff}
- Popularity: ${(puzzle.popularity / 100).toFixed(2)}
- Popularity component (pop^${beta}): ${popularityComponent.toFixed(3)}
- Rating penalty (1/(1+diff^${gamma})): ${ratingPenalty.toFixed(3)}
- Final score: ${score.toFixed(3)}
`);
    
    return score;
}

function pickPuzzleForTheme(
    theme: string,
    puzzles: Puzzle[],
    userThemeRatings: UserThemeRatings,
    usedPuzzleIds: Set<string>
): Puzzle | null {
    log(`\n=== Picking puzzle for theme: ${theme} ===`);
    const userRating = userThemeRatings[theme] || DEFAULT_RATING;  // Use default rating if theme rating doesn't exist
    log(`User rating for theme: ${userRating} ${userThemeRatings[theme] ? '' : '(default)'}`);

    // First filter: theme and unused status
    const themeFilteredPuzzles = puzzles.filter(p => 
        !usedPuzzleIds.has(p.id) && p.themes.includes(theme)
    );
    log(`Found ${themeFilteredPuzzles.length} unused puzzles for theme ${theme}`);

    // Second filter: rating range (using strict ±150 range)
    const ratingFilteredPuzzles = themeFilteredPuzzles.filter(p => 
        Math.abs(p.rating - userRating) <= RATING_RANGE
    );
    log(`Found ${ratingFilteredPuzzles.length} puzzles within rating range ±${RATING_RANGE} of user rating ${userRating}`);

    if (ratingFilteredPuzzles.length === 0) {
        // If no puzzles in range, take the closest ones but log a warning
        themeFilteredPuzzles.sort((a, b) => 
            Math.abs(a.rating - userRating) - Math.abs(b.rating - userRating)
        );
        const closestPuzzles = themeFilteredPuzzles.slice(0, MAX_PUZZLES_TO_SCORE);
        log(`WARNING: No puzzles within ideal range (±${RATING_RANGE}). Using ${closestPuzzles.length} closest puzzles instead.`);
        log(`Rating range of closest puzzles: ${Math.min(...closestPuzzles.map(p => p.rating))} to ${Math.max(...closestPuzzles.map(p => p.rating))}`);
        return selectPuzzleFromPool(closestPuzzles, userThemeRatings);
    }

    // If we have too many puzzles in range, take a random sample
    let puzzlesToScore = ratingFilteredPuzzles;
    if (ratingFilteredPuzzles.length > MAX_PUZZLES_TO_SCORE) {
        // Shuffle and take first MAX_PUZZLES_TO_SCORE
        puzzlesToScore = [...ratingFilteredPuzzles]
            .sort(() => Math.random() - 0.5)
            .slice(0, MAX_PUZZLES_TO_SCORE);
        log(`Randomly selected ${MAX_PUZZLES_TO_SCORE} puzzles from ${ratingFilteredPuzzles.length} in range`);
    }

    return selectPuzzleFromPool(puzzlesToScore, userThemeRatings);
}

function selectPuzzleFromPool(
    puzzles: Puzzle[],
    userThemeRatings: UserThemeRatings
): Puzzle | null {
    if (puzzles.length === 0) {
        log(`No valid puzzles found`);
        return null;
    }

    // Log distribution of available puzzles
    const ratings = puzzles.map(p => p.rating);
    log(`
Available puzzle statistics:
- Count: ${puzzles.length}
- Rating range: ${Math.min(...ratings)} to ${Math.max(...ratings)}
`);

    // Compute scores for puzzles
    const puzzleScores = puzzles.map(puzzle => ({
        puzzle,
        score: computePuzzleScore(puzzle, userThemeRatings)
    }));

    // Sort by score for logging
    puzzleScores.sort((a, b) => b.score - a.score);
    log('\nTop 5 puzzle scores:');
    puzzleScores.slice(0, 5).forEach(({puzzle, score}) => {
        log(`Puzzle ${puzzle.id}: score=${score.toFixed(3)}, rating=${puzzle.rating}, diff=${Math.abs(puzzle.rating - userThemeRatings[puzzle.themes[0]])}`);
    });

    // Weighted random selection
    const totalScore = puzzleScores.reduce((sum, {score}) => sum + score, 0);
    const rand = Math.random() * totalScore;
    let cumSum = 0;
    let selectedPuzzle: Puzzle | null = null;

    for (const {puzzle, score} of puzzleScores) {
        cumSum += score;
        if (rand <= cumSum) {
            selectedPuzzle = puzzle;
            break;
        }
    }

    // Fallback to highest scored puzzle if random selection fails
    if (!selectedPuzzle) {
        selectedPuzzle = puzzleScores[0].puzzle;
    }

    log(`\n=== Final Selection ===`);
    log(`Selected puzzle ${selectedPuzzle.id}:
- Rating: ${selectedPuzzle.rating}
- Themes: ${selectedPuzzle.themes.join(', ')}
- Popularity: ${(selectedPuzzle.popularity / 100).toFixed(2)}
- Score: ${puzzleScores.find(ps => ps.puzzle.id === selectedPuzzle!.id)!.score.toFixed(3)}
- Rating difference from user: ${Math.abs(selectedPuzzle.rating - userThemeRatings[selectedPuzzle.themes[0]])}
`);

    return selectedPuzzle;
}

export function getNextPuzzle(
    userThemeRatings: UserThemeRatings,
    puzzles: Puzzle[],
    usedPuzzleIds: Set<string>
): Puzzle | null {
    log('Starting puzzle selection process');
    log(`User has ratings for ${Object.keys(userThemeRatings).length} themes`);
    
    // Compute theme probabilities and select theme
    const themeProbs = computeThemeProbabilities(userThemeRatings);
    const selectedTheme = pickTheme(themeProbs);
    
    // Pick and return puzzle for selected theme
    return pickPuzzleForTheme(selectedTheme, puzzles, userThemeRatings, usedPuzzleIds);
} 